### PR TITLE
#161967112 Test user registration and login  functionality 

### PR DESCRIPTION
**Pull request by davishooly**
### What does this PR do?
  - Add test for authentication functionality to Authors Haven Django App
<hr>

### Description of Task to be completed?
  - Add test for user registration 
  - Add test for user login
  - Add test for creating a superuser
  - Send coverage to coveralls after successful Travis build
<hr>

### How should this be manually tested?
 - Clone the repo and cd into the project folder
 - Create a virtual environment by running `$ python3 -m venv <name_of_environment>`
 - Activate the virtual environment by running `source <name_of_environment>/bin/activate`
 - Run `$ pip install -r requirements.txt`
 - Then add the following environment variables to your computer.

       -`$ export DB_NAME='choose_a_dbname'`
       -`$ export DB_USER='db_username'`
       -`$ export DB_PASSWORD='db_role_password'`
       -`$ export DB_HOST='localhost or preffered host'`
       -`$ export DB_PORT='server_port_number'`

- Change the current branch to ch-test-initial-app-161967112
- Run `$ python manage.py test`
<hr>

### What are the relevant pivotal tracker stories?
[ #161967112 ](https://www.pivotaltracker.com/story/show/161967112)
<hr>

#### Checklist
- My code follows the style guidelines of this project
- I have performed a self-review of my own code
- I have commented my code, particularly in hard-to-understand areas
- I have made corresponding changes to the documentation
- My changes generate no new warnings

### CODE REVIEW

File: authors/apps/authentication/tests/test_config.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestConfigure, cls).setUpClass()
+        cls.client = APIClient()
+
+    def setUp(self):
+        test_data = UserFactory()
+        self.user = {
+            "user": {
```
Comment (bettblake08) [inline authors/apps/authentication/tests/test_config.py:20]: Do you need a user key in a variable named user?

File: authors/apps/authentication/tests/test_config.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+from rest_framework.test import APIClient
+from django.test import TestCase
+
+# local imports
+from .factory.auths_factory import UserFactory
+
+
+class TestConfigure(TestCase):
```
Comment (bettblake08) [inline authors/apps/authentication/tests/test_config.py]: Is it ok if you can change the TestConfigure name to TestConfiguration?

File: authors/apps/authentication/tests/test_login.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+# module import
+from django.urls import reverse
+from rest_framework import status
+
+# local import
+from .test_config import TestConfigure
+
+
+class TestLogin(TestConfigure):
+    pass
```
Comment (bettblake08) [inline authors/apps/authentication/tests/test_login.py]: There is no need for `pass`.

File: authors/apps/authentication/tests/test_login.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+    def test_incorrect_password(self):
+        """ test incorrect password login """
+
+        self.client.post(
+            reverse("create"), self.user,
```
Comment (bettblake08) [inline authors/apps/authentication/tests/test_login.py]: You need to update the hanging indentations to conform to Python standard. Stack them on top of each other.

File: authors/apps/authentication/tests/test_login.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            reverse("login"),
+            self.user,
+            content_type='application/json'
+        )
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_login_email_does_not_exist(self):
+        """ test login user does not exist """
+        self.client.post(
+            reverse("create"), self.user,
```
Comment (bettblake08) [inline authors/apps/authentication/tests/test_login.py]: Incorrect hanging indentations. Please refer to the previous comment.

File: authors/apps/authentication/urls.py [hunk 6]
```diff
@@ -5,7 +5,7 @@
 )
 
 urlpatterns = [
-    url(r'^user/?$', UserRetrieveUpdateAPIView.as_view()),
-    url(r'^users/?$', RegistrationAPIView.as_view()),
-    url(r'^users/login/?$', LoginAPIView.as_view()),
+    url(r'^user/?$', UserRetrieveUpdateAPIView.as_view(), name="update"),
```
Comment (bettblake08) [inline authors/apps/authentication/urls.py]: The urls name are not endpoint specific enough. Eg. change update url to read_and_update_user.

File: authors/apps/authentication/tests/test_register.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            content_type='application/json'
+        )
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(
+            response.data,
+            {'email': f'{self.user["user"]["email"]}',
+             'username': f'{self.user["user"]["username"]}'})
+
+    def test_create_super_user(self):
```
Comment (bettblake08) [inline authors/apps/authentication/tests/test_register.py]: Testing for superuser should be done in a separate test file. It's not part of the register user API endpoint tests.

**Review by bettblake08 — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/tests/factory/auths_factory.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+
+class UserFactory(factory.Factory):
+
+    class Meta:
+        model = User
+
+    username = factory.Sequence(lambda n: "kimame%d" % n)
+    email = factory.LazyAttribute(lambda b: "%s@gmail.com" % b.username)
+    password = "password"
```
Comment (denomash) [inline authors/apps/authentication/tests/factory/auths_factory.py:13]: include comments in this file for proper documentation.

File: authors/apps/authentication/tests/test_config.py (hunk 1 above)
Comment (denomash) [reply authors/apps/authentication/tests/test_config.py:20]: add comments in this file too

Comment (davishooly) [reply authors/apps/authentication/tests/test_config.py:20]: I think there can be another way to implement user data without having the user key but for our case, we followed the conventions provided in the ReadMe file

**Review by denomash — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/tests/test_login.py (hunk 3 above)
Comment (davishooly) [reply authors/apps/authentication/tests/test_login.py]: nice I didn't see that

File: authors/apps/authentication/tests/test_register.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
+# module import
+from rest_framework import status
+import json
```
Comment (Bryan-Cee) [inline authors/apps/authentication/tests/test_register.py]: This import has not been used, please erase it.

**Review by Bryan-Cee — CHANGES_REQUESTED** [review]

File: authors/apps/authentication/tests/test_login.py (hunk 4 above)
Comment (davishooly) [reply authors/apps/authentication/tests/test_login.py]: I,ll need to configure flex8

**Review by denomash — APPROVED** [review]

**Review by Bryan-Cee — APPROVED** [review]
👍

**Review by bettblake08 — APPROVED** [review]
👍 Its all good.

**Review by Darius-Ndubi — APPROVED** [review]
👍

**Review by mutoromike — APPROVED** [review]
LGTM

**Pull request merged**